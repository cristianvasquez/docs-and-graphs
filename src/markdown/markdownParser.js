import remarkFrontmatter from 'remark-frontmatter'
import remarkGfm from 'remark-gfm'
import remarkParse from 'remark-parse'
import wikiLinkPlugin from 'remark-wiki-link'
import { unified } from 'unified'

function createMarkdownParser () {
  const parser = unified().
    use(remarkParse).
    use(remarkFrontmatter).
    use(remarkGfm).
    use(wikiLinkPlugin)
  return {
    parse: async (string) => {
      return parser.parse(string)
    },
  }
}

export { createMarkdownParser }
