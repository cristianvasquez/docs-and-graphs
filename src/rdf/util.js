function pushToList (pointer, predicate, term) {
  if (pointer.out(predicate).isList()) {
    const elements = [...pointer.out(predicate).list(), term]
    pointer.deleteList(predicate)
    pointer.addList(predicate, elements)
  } else {
    pointer.addList(predicate, [term])
  }
}

export { pushToList }
