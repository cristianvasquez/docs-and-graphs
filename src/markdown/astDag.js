import rdf from 'rdf-ext'
import ns from '../namespaces.js'
import { getList, getText } from './ast.js'
import {extractTags} from '../text/tags.js'

async function getAstDag ({ astNode, fullText }) {

  const dataset = rdf.dataset()

  const root = rdf.blankNode()
  const pointer = rdf.clownface({ term: root, dataset })
  const quads = []
  pointer.addOut(ns.rdf.type,ns.mark.Root)

  let headersStack = [{
    iri: root,
    depth: 0,
  }]

  astNode.children.reduce((currentUri, astNode) => {
    const text = getText({ astNode, fullText })

    if (astNode.type === 'heading') {
      const currentIri = rdf.blankNode()
      const ancesters = headersStack.filter(x => x.depth < astNode.depth)
      if (ancesters.length) {
        headersStack = headersStack.filter(x => x.depth < astNode.depth)
      }

      const parentIri = ancesters.length
        ? ancesters[ancesters.length - 1].iri
        : currentUri

      quads.push(rdf.quad(parentIri, ns.mark.contains, currentIri))
      quads.push(rdf.quad(currentIri, ns.mark.header, rdf.literal(text)))

      for (const tag of extractTags(text)){
        quads.push(rdf.quad(currentIri, ns.mark.tag, rdf.literal(tag)))
      }

      headersStack.push({
        iri: currentIri,
        depth: astNode.depth,
      })
      return currentIri
    } else {
      if (astNode.type === 'list') {
        for (const text of getList({ astNode: astNode, fullText: fullText })) {
          quads.push(rdf.quad(currentUri, ns.mark.listItem, rdf.literal(text)))
        }
      } else if (astNode.type === 'paragraph') {
        quads.push(rdf.quad(currentUri, ns.mark.item, rdf.literal(text)))
      }
    }
    return currentUri
  }, root)
  pointer.dataset.addAll(quads)
  return pointer
}

export { getAstDag }
