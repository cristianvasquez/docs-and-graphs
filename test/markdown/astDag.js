import { expect } from 'expect'

import toMatchSnapshot from 'expect-mocha-snapshot'
import { getAstDag } from '../../src/markdown/astDag.js'
import { createMarkdownParser } from '../../src/markdown/markdownParser.js'
import { prettyPrint } from '../util.js'
import tests from '../tests.js'

expect.extend({ toMatchSnapshot })

describe('astDag', async function () {
  for (const current of tests) {
    it(current.title, async function () {
      const fullText = current.markdown
      const parser = createMarkdownParser()
      const astNode = await parser.parse(fullText)

      const astGraph = await getAstDag({ astNode, fullText })

      const turtle = await prettyPrint(astGraph.dataset)
      expect(turtle).toMatchSnapshot(this)
    })
  }
})
