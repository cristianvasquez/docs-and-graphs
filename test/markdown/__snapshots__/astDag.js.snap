// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`astDag can use a taxonomy (recursive patterns) 1`] = `
"@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .
@prefix purl: <http://purl.org/dc/terms/> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org/> .
@prefix dot: <http://dottriples.org/> .
@prefix mark: <http://markdown.org/> .

_:b21 a mark:Root ;
	mark:contains [
		mark:contains [
			mark:contains [
				mark:contains [
					mark:header \\"#### Tomato\\" ;
					mark:item \\"A tomato most of the time is red in Europe\\" ;
				] ;
				mark:header \\"### Red\\" ;
			], [
				mark:header \\"### Green\\" ;
				mark:listItem \\"- Lettuce\\" ;
			] ;
			mark:header \\"## Food\\" ;
		], [
			mark:contains [
				mark:header \\"### Metal\\" ;
				mark:listItem \\"* Screwdriver\\", \\"* Hammer\\" ;
			], [
				mark:header \\"### Wood\\" ;
				mark:listItem \\"- 1\\", \\"- 2\\", \\"- 3\\" ;
			] ;
			mark:header \\"## Tools\\" ;
		] ;
		mark:header \\"# Things\\" ;
	] .

"
`;

exports[`astDag produces quads using embedded tag 1`] = `
"@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .
@prefix purl: <http://purl.org/dc/terms/> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org/> .
@prefix dot: <http://dottriples.org/> .
@prefix mark: <http://markdown.org/> .

_:b1 a mark:Root ;
	mark:contains [
		mark:contains [
			mark:contains [
				mark:header \\"### Who\\" ;
				mark:listItem \\"- Bob (action::prepare)\\", \\"- Alice (action::cook)\\" ;
			], [
				mark:header \\"### Purpose\\" ;
				mark:item \\"To make the best Pancakes in the world\\" ;
			], [
				mark:contains [
					mark:header \\"#### Embedded\\" ;
				] ;
				mark:header \\"### Song\\" ;
				mark:item \\"Pancakes, Pancakes !\\" ;
			], [
				mark:header \\"### Tasks\\" ;
				mark:item \\"And a paragraph, explaining even more things\\" ;
				mark:listItem \\"- Mix ingredients\\", \\"- Heat the pan\\", \\"- Prepare the Pancakes\\" ;
			] ;
			mark:header \\"## Pancake maker #role\\" ;
			mark:item \\"Some text for the Pancake maker\\" ;
			mark:tag \\"#role\\" ;
		], [
			mark:contains [
				mark:header \\"### Who\\" ;
				mark:listItem \\"- Charlie\\" ;
			], [
				mark:header \\"### Purpose\\" ;
				mark:item \\"Sell the Pancakes to the customers\\" ;
			], [
				mark:header \\"### Tasks\\" ;
				mark:listItem \\"- Coordination of Pancake economic activities\\", \\"- Advertise the Pancakes\\" ;
			] ;
			mark:header \\"## Pancake seller #role\\" ;
			mark:tag \\"#role\\" ;
		] ;
		mark:header \\"# Roles\\" ;
		mark:item \\"The roles are the following: \\" ;
	], [
		mark:header \\"# Some other topic\\" ;
		mark:item \\"With a description\\" ;
	] .

"
`;

exports[`astDag produces quads using lists 1`] = `
"@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .
@prefix purl: <http://purl.org/dc/terms/> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org/> .
@prefix dot: <http://dottriples.org/> .
@prefix mark: <http://markdown.org/> .

_:b14 a mark:Root ;
	mark:contains [
		mark:contains [
			mark:header \\"## Properties\\" ;
			mark:item \\"age :: 22\\" ;
		] ;
		mark:header \\"# Alice\\" ;
	], [
		mark:contains [
			mark:header \\"## Interests\\" ;
			mark:listItem \\"- Eating\\" ;
		] ;
		mark:header \\"# Bob (age :: 42)\\" ;
	], [
		mark:contains [
			mark:header \\"## Description\\" ;
			mark:item \\"Charlie is a good guy. (age :: 56)\\" ;
		] ;
		mark:header \\"# Charlie\\" ;
	] .

"
`;
